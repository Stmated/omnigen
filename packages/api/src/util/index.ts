export * from './Environment';

export type Arrayable<T> = T | Array<T>;
export type ReadonlyArrayable<T> = T | ReadonlyArray<T>;
