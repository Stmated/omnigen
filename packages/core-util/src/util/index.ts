
export * from './ProtocolHandler.js';
export * from './Sorters.js';
export * from './ToString.js';
export * from './Case.js';
export * from './Util.ts';

export function assertUnreachable(x: never): never {
  throw new Error('Unreachable code was reached!');
}

export function assertDefined<T>(x: T): Exclude<T, undefined | null | void> {
  if (x === undefined || x === null) {
    throw new Error(`A reduced node became 'undefined' when it is required.`);
  }

  // @ts-ignore
  return x;
}

export function isDefined<T>(argument: T | undefined): argument is T {
  return argument !== undefined;
}
