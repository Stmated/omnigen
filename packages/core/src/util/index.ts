import {Simplify, Something} from './TsTypes.ts';

export * from './ProtocolHandler';
export * from './Sorters';
export * from './ToString';
export * from './Case';
export * from './Util';
export * from './CombineTypeUtils';
export * from './TsTypes';

export * from './assertions.ts';

export function assertUnreachable(x: never): never {
  throw new Error(`Unreachable code was reached, with: ${getShallowPayloadString(x)}`);
}

export function assertDefined<T>(x: T): Exclude<T, undefined | null | void> {
  if (x === undefined || x === null) {
    throw new Error(`A reduced node became 'undefined' when it is required.`);
  }

  // @ts-ignore
  return x;
}

export function isDefined<T>(argument: T | undefined | void): argument is T {
  return argument !== undefined;
}

export function getShallowPayloadString<T>(origin: T, maxDepth = 1) {
  return getShallowPayloadStringInternal(origin, 0, maxDepth);
}

export function getShallowPayloadStringInternal<T>(origin: T, depth: number, maxDepth: number): string {

  return JSON.stringify(origin, (key, value) => {
    if (value && typeof value === 'object' && key) {
      if (depth < maxDepth) {
        return JSON.parse(getShallowPayloadStringInternal(value, depth + 1, maxDepth));
      }
      return '[...]';
    }
    return value;
  });
}

function expectToBeInstanceOf<const T, const C extends new(...args: any[]) => T>(value: T, clss: C): asserts value is InstanceType<C> {
  if (!(value instanceof clss)) {
    throw new Error(`Expected value to be an instance of ${clss.name}`);
  }
}

export function expectToBeDefined<T>(val?: T): asserts val is Simplify<Something<T>> {
  if (!val) {
    throw new Error(`Expected value ${val} to be defined`);
  }
}

export function expectPropertyToBe<O, K extends keyof O, const V extends O[K]>(owner: O, key: K, compareVal: V): asserts owner is Extract<O, Record<K, V>> {
  if (owner[key] !== compareVal) {
    throw new Error(`${owner[key]} should have been ${compareVal}`);
  }
}

export interface ExpectTs {
  toBeDefined: typeof expectToBeDefined,
  toBeInstanceOf: typeof expectToBeInstanceOf,
  propertyToBe: typeof expectPropertyToBe,
}

export const expectTs: ExpectTs = {
  toBeDefined: expectToBeDefined,
  toBeInstanceOf: expectToBeInstanceOf,
  propertyToBe: expectPropertyToBe,
} as const;
