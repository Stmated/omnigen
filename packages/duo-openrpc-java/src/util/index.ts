
export * from './JavaTestUtils.ts';
export * from './OpenRpcTestUtils.ts';
