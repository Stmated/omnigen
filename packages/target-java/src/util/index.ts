export * from './JavaUtil';
export * from './FreeTextUtils';
