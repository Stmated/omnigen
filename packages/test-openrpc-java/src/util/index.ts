
export * from './JavaTestUtils.ts';
export * from './OpenRpcTestUtils.ts';
export * from './ParsedJavaTestVisitor.ts';
