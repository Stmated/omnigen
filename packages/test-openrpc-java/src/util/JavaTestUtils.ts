import * as JavaParser from 'java-parser';
import {ModelTransformOptions, PackageOptions, RenderedCompilationUnit} from '@omnigen/api';
import {JavaOptions, SerializationLibrary, ZodJavaOptions} from '@omnigen/target-java';
import {ParsedJavaTestVisitor} from '../util';
import {DEFAULT_SPECIFIC_TEST_TARGET_OPTIONS} from '@omnigen/utils-test';
import {Util, ZodCompilationUnitsContext} from '@omnigen/core';
import {PluginManager} from '@omnigen/plugin';
import {BaseContext, FileContext, TargetContext} from '@omnigen/core-plugin';
import {z, ZodObject} from 'zod';
import {LoggerFactory} from '@omnigen/core-log';
import {OpenRpcPlugins} from '@omnigen/parser-openrpc';

const logger = LoggerFactory.create(import.meta.url);

export const DEFAULT_SPECIFIC_TEST_JAVA_OPTIONS: Partial<JavaOptions> = {
  serializationLibrary: SerializationLibrary.POJO,
  beanValidation: false,
};

export const DEFAULT_TEST_JAVA_OPTIONS: JavaOptions = {
  ...ZodJavaOptions.parse({}),
  ...DEFAULT_SPECIFIC_TEST_TARGET_OPTIONS,
  ...DEFAULT_SPECIFIC_TEST_JAVA_OPTIONS,
};

export interface JavaTestUtilsOptions {
  // parserOptions?: Partial<ParserOptions>,
  modelTransformOptions?: Partial<ModelTransformOptions>,
  // targetOptions?: Partial<TargetOptions>,
  packageOptions?: Partial<PackageOptions>,
  javaOptions?: Partial<JavaOptions>,
  arguments?: Record<string, any>,
}

export class JavaTestUtils {

  constructor() {
    logger.debug(`Loaded OpenRPC: ${OpenRpcPlugins.OpenRpcPlugin}`);
  }

  public static async getFileContentsFromFile(fileName: string, options?: JavaTestUtilsOptions, source = 'openrpc'): Promise<Map<string, string>> {

    const filePath = Util.getPathFromRoot(`./packages/parser-${source}/examples/${fileName}`);
    const result = await JavaTestUtils.getResultFromFilePath(filePath, options || {}, ZodCompilationUnitsContext);

    return JavaTestUtils.cuToContentMap(result.compilationUnits);
  }

  public static async getResultFromFilePath<
    E extends ZodObject<any>,
    S extends ZodObject<any>
  >(
    filePath: string,
    options: JavaTestUtilsOptions,
    expected: E,
    stopAt?: S,
  ): Promise<z.output<E>> {

    const ctx: BaseContext & FileContext & TargetContext = {
      file: filePath,
      // target: 'java', // <-- implicit, because we have no other target plugin in this package
      defaults: {
        ...DEFAULT_SPECIFIC_TEST_TARGET_OPTIONS,
        ...DEFAULT_SPECIFIC_TEST_JAVA_OPTIONS,
      },
      arguments: {
        ...(options.modelTransformOptions),
        ...(options.packageOptions),
        ...(options.javaOptions),
        ...(options.arguments),
      },
    };

    const pm = new PluginManager({includeAuto: true});
    const result = await pm.execute({ctx: ctx, debug: true, stopAt: stopAt});
    const last = result.results[result.results.length - 1];

    return expected.parse(last.ctx) as z.output<E>;
  }

  public static getParsedContent(fileContents: Map<string, string>, fileName: string): ParsedJavaTestVisitor {

    const content = fileContents.get(fileName) || '';
    let cst: JavaParser.CstNode;
    try {
      cst = JavaParser.parse(content);
      if (!cst) {
        throw new Error(`The JavaParser must return something`);
      }

    } catch (ex) {
      throw new Error(`Could not parse '${fileName}'': ${ex}:\n${content}`, {cause: ex instanceof Error ? ex : undefined});
    }

    const visitor = new ParsedJavaTestVisitor();
    visitor.visit(cst);

    return visitor;
  }

  public static cuToContentMap(compilationUnits: RenderedCompilationUnit[]) {
    const fileContents = new Map<string, string>();

    for (const render of compilationUnits) {
      fileContents.set(render.fileName, render.content);
    }

    return fileContents;
  }
}
