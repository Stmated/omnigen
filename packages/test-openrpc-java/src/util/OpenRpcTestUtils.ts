import fs from 'fs/promises';
import {JavaTestUtils, JavaTestUtilsOptions} from './JavaTestUtils.ts';
import {OmniModelParserResult, PackageOptions, TargetOptions} from '@omnigen/api';
import {JavaPlugins, JavaOptions} from '@omnigen/target-java';
import {ZodModelContext, ZodPackageOptionsContext, ZodTargetOptionsContext} from '@omnigen/core-plugin';
import {Util, ZodCompilationUnitsContext} from '@omnigen/core';

export type KnownSchemaNames = 'openrpc';

/**
 * TODO: REMOVE THIS WHOLE THING! OR AT LEAST SERIOUSLY MINIMIZE IT! Use `Omnigen` base class instead!
 *        `Omnigen` needs to be made more dynamic/non-hardcoded first, then this class can just help with some basic options/args!
 */
export class OpenRpcTestUtils {

  static getKnownSchemaNames(): KnownSchemaNames[] {
    return ['openrpc'];
  }

  static async listExampleFileNames(type: KnownSchemaNames): Promise<string[]> {
    const dirPath = Util.getPathFromRoot(`./packages/parser-${type}/examples/`);
    return fs.readdir(dirPath, {withFileTypes: true})
      .then(paths => {
        return paths.filter(it => it.isFile()).map(it => it.name);
      });
  }

  /**
   * TODO: Need a way to state what the "final" plugin should be
   *        since we should NOT render things here, we only care about the generated model
   */
  static async readExample(
    type: KnownSchemaNames,
    fileName: string,
    options: JavaTestUtilsOptions,
  ): Promise<OmniModelParserResult<JavaOptions & PackageOptions & TargetOptions>> {

    const result = await JavaTestUtils.getResultFromFilePath(
      Util.getPathFromRoot(`./packages/parser-${type}/examples/${fileName}`),
      options,
      ZodModelContext
        .merge(JavaPlugins.ZodJavaOptionsContext)
        .merge(ZodPackageOptionsContext)
        .merge(ZodTargetOptionsContext),
      ZodCompilationUnitsContext,
    );

    return {
      model: result.model,
      options: {
        ...result.javaOptions,
        ...result.packageOptions,
        ...result.targetOptions,
      },
    };
  }
}
