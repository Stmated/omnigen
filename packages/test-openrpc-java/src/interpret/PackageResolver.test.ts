import {OmniKindPrimitive, PackageOptions} from '@omnigen/api';
import {JavaTestUtils} from '../util';
import {JavaOptions, SerializationLibrary} from '@omnigen/target-java';
import {describe, test, vi} from 'vitest';

describe('PackageResolver', () => {

  test.concurrent('FromSchema', async ctx => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const fileContents = await JavaTestUtils.getFileContentsFromFile('packages.json', {
      modelTransformOptions: {generifyTypes: false},
      javaOptions: {preferNumberType: OmniKindPrimitive.DOUBLE},
    });

    ctx.expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      ctx.expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${ctx.task.suite?.name}/${ctx.task.name}/${fileName}`);
    }
  });

  test.concurrent('FromCode', async ctx => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const javaOptions: Partial<JavaOptions> = {
      includeGenerated: false,
      serializationLibrary: SerializationLibrary.POJO,
      additionalPropertiesInterfaceAfterDuplicateCount: 1,
      allowCompressInterfaceToInner: false,
    };

    const packageOptions: Partial<PackageOptions> = {
      packageResolver: (_type, typeName) => {
        if (typeName.match(/.*Error.*/i)) {
          return 'some.base.pkg.errors';
        }
        if (typeName.match(/JsonRpc.*/i)) {
          return 'some.base.pkg';
        }
        return 'some.other.pkg';
      },
    };

    const fileContents = await JavaTestUtils.getFileContentsFromFile('additional-properties.json', {
      javaOptions: javaOptions,
      packageOptions: packageOptions,
    });

    ctx.expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      ctx.expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${ctx.task.suite?.name}/${ctx.task.name}/${fileName}`);
    }
  });
});
