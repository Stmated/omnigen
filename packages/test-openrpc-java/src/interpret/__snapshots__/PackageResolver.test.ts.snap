// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PackageResolver > FromCode 1`] = `
[
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "IAdditionalProperties.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
  "ListThingsRequest.java",
  "ListThingsRequestParams.java",
  "ListThingsResponse.java",
  "Thing.java",
]
`;

exports[`PackageResolver > FromSchema 1`] = `
[
  "A.java",
  "Abs.java",
  "B.java",
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "GiveIn1GetOut1Request.java",
  "GiveIn1GetOut1RequestParams.java",
  "GiveIn1GetOut1Response.java",
  "GiveIn2GetOut2Request.java",
  "GiveIn2GetOut2RequestParams.java",
  "GiveIn2GetOut2Response.java",
  "In1.java",
  "In2.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
]
`;
